_posts/2017-06-08-Rust-1.18.md

```diff
@@ -0,0 +1,200 @@
+`bar` would be public to the entire crate, but not outside of it. This makes it
+easier to declare APIs that are "public to your crate", but not exposed to your
+users. This was *possible* with the current module system, but often very awkward.
+
+You can also specify a path, like this:
+
+```rust
+pub(a::b::c) foo;
+```
+
+This means "usable within the hierarchy of `a::b::c`, but not elsewhere." This
+feature was defined in [RFC
+1422](https://github.com/rust-lang/rfcs/blob/master/text/1422-pub-restricted.md)
+
+For our Windows users, Rust 1.18.0 has [a new attribute,
+`#![windows_subsystem]`](https://github.com/rust-lang/rust/pull/40870). It
+works like this:
+
+```rust
+#![windows_subsystem(console)]
+#![windows_subsystem(windows)]
+```
+
+These control the [`/SUBSYSTEM` flag](https://msdn.microsoft.com/en-us/library/fcc1zstk.aspx)
+in the linker. For now, only `console` and `windows` are supported.
+
+When is this useful? In the simplest terms, if you're developing a graphical
+application, and do not specify `windows`, a console window would flash up upon
+your application's start. With this flag, it won't.
+
+Finally, Rust's tuples, enum variant fields, and structs (without `#[repr]`) have
+always had an undefined layout. [We've turned on automatic re-ordering](https://github.com/rust-lang/rust/pull/40377), which can result in smaller sizes
+through reducing padding. Consider a struct like this:
+
+```rust
+struct Suboptimal(u8, u16, u8);
+```
+
+In previous versions of Rust on the x86_64 platform, this struct would have the
+size of six bytes. But looking at the source, you'd expect it to have four. The
+extra two bytes come from padding; given that we have a `u16` here, it should be
+aligned to two bytes. But in this case, it's at offset one. To move it to offset
+two, another byte of padding is placed after the first `u8`. To give the whole struct
+a proper alignment, another byte is added after the second `u8` as well, giving us
+`1 + 1 (padding) + 2 + 1 + 1 (padding) = 6 bytes`.
+
+But what if our struct looked like this?
+
+```rust
+struct Optimal(u8, u8, u16);
+```
+
+This struct is properly aligned; the `u16` lies on a two byte boundary, and so does
+the entire struct. No padding is needed. This gives us `1 + 1 + 2 = 4 bytes`.
+
+When designing Rust, stuff like this is why we left the details here undefined.
+It allows for exactly this kind of thing: the compiler can optimize `Suboptimal`
+into `Optimal` automatically. And if you check the sizes of `Suboptimal` and
+`Optimal` on Rust 1.18.0, you'll see that they both have a size of four bytes.
+
+We've been planning this for a while; previous versions of Rust included this
+optimization on the nightly channel, but some people wrote unsafe code that
+assumed the exact details of the representation. We rolled it back while we fixed
+all instances of this that we know about, but if you find some code breaks due to
```

Comment:
What kind of code would break because of this? What does breaking look like? How do I fix it? This is a little scary to me without it, even though I probably don't have this problem ;)

Is there someplace we can link that would go into more details on the potential breakage?